File: behavioral/dart/mediator_structured/smart_home.dart
import 'interfaces.dart';

class SmartDevice {
  final Mediator _homeHub;
  final String _deviceName;
  bool _isOn = false;
  
  SmartDevice(this._homeHub, this._deviceName);
  
  void turnOn() {
    _isOn = true;
    print('$_deviceName turned ON');
    _homeHub.sendMessage('$_deviceName turned ON', _deviceName);
  }
  
  void turnOff() {
    _isOn = false;
    print('$_deviceName turned OFF');
    _homeHub.sendMessage('$_deviceName turned OFF', _deviceName);
  }
  
  String get name => _deviceName;
  
  bool get isOn => _isOn;
}

class SmartLight extends SmartDevice {
  SmartLight(super.homeHub, super.name);
}

class SmartThermostat extends SmartDevice {
  int _temperature = 70;
  
  SmartThermostat(super.homeHub, super.name);
  
  void setTemperature(int temp) {
    _temperature = temp;
    print('$name set to $_temperature°F');
  }
  
  int get temperature => _temperature;
}

class HomeHub implements Mediator {
  final Map<String, SmartDevice> _devices = {};
  
  void addDevice(SmartDevice device) {
    _devices[device.name] = device;
  }
  
  @override
  void sendMessage(String message, String sender) {
    print('Home Hub: $message');
    
    if (sender.contains('Thermostat') && message.contains('turned ON')) {
      for (final device in _devices.values) {
        if (device.name.contains('Light') && !device.isOn) {
          device.turnOn();
        }
      }
    }
  }
  
  void activateAwayMode() {
    print('Home Hub: Activating away mode...');
    for (final device in _devices.values) {
      if (device.isOn) {
        device.turnOff();
      }
    }
  }
  
  void activateHomeMode() {
    print('Home Hub: Activating home mode...');
    for (final device in _devices.values) {
      if (device.name.contains('Light')) {
        device.turnOn();
      }
    }
  }
}

File: behavioral/typescript/mediator_structured/smart_home.ts
import { Mediator } from './interfaces';

export class SmartDevice {
  protected homeHub: Mediator;
  protected deviceName: string;
  protected isOn: boolean = false;
  
  constructor(hub: Mediator, name: string) {
    this.homeHub = hub;
    this.deviceName = name;
  }
  
  turnOn(): void {
    this.isOn = true;
    console.log(`${this.deviceName} turned ON`);
    this.homeHub.sendMessage(`${this.deviceName} turned ON`, this.deviceName);
  }
  
  turnOff(): void {
    this.isOn = false;
    console.log(`${this.deviceName} turned OFF`);
    this.homeHub.sendMessage(`${this.deviceName} turned OFF`, this.deviceName);
  }
  
  getName(): string {
    return this.deviceName;
  }
  
  getStatus(): boolean {
    return this.isOn;
  }
}

export class SmartLight extends SmartDevice {
  constructor(hub: Mediator, name: string) {
    super(hub, name);
  }
}

export class SmartThermostat extends SmartDevice {
  private temperature: number = 70;
  
  constructor(hub: Mediator, name: string) {
    super(hub, name);
  }
  
  setTemperature(temp: number): void {
    this.temperature = temp;
    console.log(`${this.getName()} set to ${this.temperature}°F`);
  }
  
  getTemperature(): number {
    return this.temperature;
  }
}

export class HomeHub implements Mediator {
  private devices: Map<string, SmartDevice> = new Map();
  
  addDevice(device: SmartDevice): void {
    this.devices.set(device.getName(), device);
  }
  
  sendMessage(message: string, sender: string): void {
    console.log(`Home Hub: ${message}`);
    
    if (sender.includes('Thermostat') && message.includes('turned ON')) {
      for (const [name, device] of this.devices) {
        if (name.includes('Light') && !device.getStatus()) {
          device.turnOn();
        }
      }
    }
  }
  
  activateAwayMode(): void {
    console.log('Home Hub: Activating away mode...');
    for (const [name, device] of this.devices) {
      if (device.getStatus()) {
        device.turnOff();
      }
    }
  }
  
  activateHomeMode(): void {
    console.log('Home Hub: Activating home mode...');
    for (const [name, device] of this.devices) {
      if (name.includes('Light')) {
        device.turnOn();
      }
    }
  }
}
